bridge-fungible-token/src/utils.sw

```diff
@@ -1,10 +1,145 @@
 const GTF_INPUT_MESSAGE_DATA_LENGTH = 0x11B;
 const GTF_INPUT_MESSAGE_DATA = 0x11E;
+const GTF_INPUT_MESSAGE_SENDER = 0x115;
+const GTF_INPUT_MESSAGE_RECIPIENT = 0x116;
+
+/// Make any necessary adjustments to decimals(precision) on the amount
+/// to be withdrawn. This amount needs to be passed via message.data as a b256
+pub fn adjust_withdrawal_decimals(val: u64) -> b256 {
+    if DECIMALS < LAYER_1_DECIMALS {
+        let amount = ~U256::from(0, 0, 0, val);
+        let factor = ~U256::from(0, 0, 0, 10.pow(LAYER_1_DECIMALS - DECIMALS));
+        let components = amount.multiply(factor).into();
+        compose(components)
+    } else {
+        // Either decimals are the same, or decimals are negative.
+        // TODO: Decide how to handle negative decimals before mainnet.
+        // For now we make no decimal adjustment for either case.
+        compose((0, 0, 0, val))
+    }
+}
+
+/// Make any necessary adjustments to decimals(precision) on the deposited value, and return either a converted u64 or an error if the conversion can't be achieved without overflow or loss of precision.
+pub fn adjust_deposit_decimals(msg_val: b256) -> Result<u64, BridgeFungibleTokenError> {
+    let decomposed = decompose(msg_val);
+    let value = ~U256::from(decomposed.0, decomposed.1, decomposed.2, decomposed.3);
+
+    if LAYER_1_DECIMALS > DECIMALS {
+        let adjustment_factor = ~U256::from(0, 0, 0, 10.pow(LAYER_1_DECIMALS - DECIMALS));
+
+        if value.divide(adjustment_factor).multiply(adjustment_factor) == value
+            && (value.gt(adjustment_factor)
+            || value.eq(adjustment_factor))
+        {
+            let adjusted = value.divide(adjustment_factor);
+            let val_result = adjusted.as_u64();
+            match val_result {
+                Result::Err(e) => {
+                    Result::Err(BridgeFungibleTokenError::BridgedValueIncompatability)
+                },
+                Result::Ok(val) => {
+                    Result::Ok(val)
+                },
+            }
+        } else {
+            Result::Err(BridgeFungibleTokenError::BridgedValueIncompatability)
+        }
+    } else {
+        // Either decimals are the same, or decimals are negative.
+        // TODO: Decide how to handle negative decimals before mainnet.
+        // For now we make no decimal adjustment for either case.
+        let val_result = value.as_u64();
+        match val_result {
+            Result::Err(e) => {
+                Result::Err(BridgeFungibleTokenError::BridgedValueIncompatability)
+            },
+            Result::Ok(val) => {
+                Result::Ok(val)
+            },
+        }
+    }
+}
+
+/// Build a single b256 value from a tuple of 4 u64 values.
+pub fn compose(words: (u64, u64, u64, u64)) -> b256 {
+    let res: b256 = 0x0000000000000000000000000000000000000000000000000000000000000000;
+    asm(r1: __addr_of(words), r2: __addr_of(res)) {
+        mcpi r2 r1 i32;
```

Comment:
does casting not work here like it does in the `compose` function? `asm(r1: __addr_of(words)) { r1: b256 }`

Reply:
It does! I shouldn't be surprised, but I feel like the compiler couldn't do this previously.
Nice to see 90% of the asm get pruned here.